### DISCUSSION TITLE
Use own Load Balancer instead of Web Adaptor

### DISCUSSION THREAD

**Discussion by jesa-esrich** [Esri/arcgis-powershell-dsc#488] [Q&A] [answered]
**Is it possible to use ArcGIS PowerShell DSC for the installation/configuration of an ArcGIS Enterprise Base Deployment with your own Load Balancers instead of the Web Adaptor?**

- And how would such a configuration JSON look like? Remove the Web Adaptor parameter and specify internal and external load balancer. (Load Balancer has to be configured in advance).
- And is this supported?
- As far as I know, the Cloud Builder uses ArcGIS PowerShell DSC in the background to set up an ArcGIS Enterprise Base Deployment with an V2 Azure Application Gateway in front (and no ArcGIS Web Adaptors). This is also seen as best practice for cloud deployments on Azure.

Use Case:
- Setting up a ArcGIS Enterprise in the Azure Cloud with V2 Azure Application Gateway (no ArcGIS Web Adaptor)
- Using ArcGIS PowerShell DSC (Cloud Builder cannot be used because of restricted privileges on the subscription)

**Comment by cameronkroeker** [comment, marked as answer]
Hi @jesa-esrich,

If you bring your own reverse proxy or load balancer then the ArcGIS Web Adaptor becomes an optional component of ArcGIS Enterprise (unless you will be using [Integrated Windows Authentication](https://enterprise.arcgis.com/en/portal/latest/administer/windows/use-integrated-windows-authentication-with-your-portal.htm) then its required). Therefore, it is possible to use ArcGIS PowerShell DSC for the installation/configuration/upgrade of an ArcGIS Enterprise Base Deployment with your own Load Balancer without using the ArcGIS Web Adaptor. As you mentioned remove the `ServerWebAdaptor` and `PortalWebAdaptor` roles and specify an external and/or internal load balancer (Load Balancer has to be configured in advance). 

> Note: The InternalLoadBalancer parameter is optional, and if you bring your own File Share then the FileShare role is not necessary either.

**ConfigData.Portal.ExternalLoadBalancer**: When set, this is used along with `ConfigData.PortalContext` parameter to set the [WebContextURL](https://developers.arcgis.com/rest/enterprise-administration/portal/system-properties.htm) property, and the [privatePortalURL](https://developers.arcgis.com/rest/enterprise-administration/portal/system-properties.htm) property (unless `ConfigData.Portal.InternalLoadBalancer` is set).

**ConfigData.Server.ExternalLoadBalancer**: When set, this is used along with `ConfigData.ServerContext` parameter to set the [WebContextURL](https://developers.arcgis.com/rest/enterprise-administration/server/serverproperties.htm) property. This is also used for the [Services URL](https://enterprise.arcgis.com/en/portal/latest/administer/windows/federate-an-arcgis-server-site-with-your-portal.htm) and [Administration URL](https://enterprise.arcgis.com/en/portal/latest/administer/windows/federate-an-arcgis-server-site-with-your-portal.htm) for federation (unless `ConfigData.InternalLoadBalancer` is set).

#### Here is a sample json file:

```
{
    "AllNodes": [
        {
            "NodeName": "[Machine 1 Name]",
            "Role": [
                "FileShare"
            ]
        },
        {
            "NodeName": "[Machine 2 Name]",
            "DataStoreTypes": [
                "Relational",
                "TileCache"
            ],
            "Role": [
                "DataStore"
            ]
        },
        {
            "NodeName": "[Machine 3 Name]",
            "DataStoreTypes": [
                "Relational",
                "TileCache"
            ],
            "Role": [
                "DataStore"
            ]
        },
        {
            "NodeName": "[Machine 4 Name]",
            "Role": [
                "Portal"
            ]
        },
        {
            "NodeName": "[Machine 5 Name]",
            "Role": [
                "Portal"
            ]
        },
        {
            "NodeName": "[Machine 6 Name]",
            "Role": [
                "Server"
            ]
        },
        {
            "NodeName": "[Machine 7 Name]",
            "Role": [
                "Server"
            ]
        }
    ],
    "ConfigData": {
        "Version": "[Version of ArcGIS Deployment]",
        "ServerContext": "server",
        "PortalContext": "portal",
        "FileShareLocalPath": "[File Share Local Path]",
        "FileShareName": "[File Share Name]",
        "ServerRole": "GeneralPurposeServer",
        "Credentials": {
            "ServiceAccount": {
                "Password": "[ServiceAccount Password]",
                "UserName": "[ServiceAccount Username - Can be a Domain Account]",
                "IsDomainAccount": false,
                "IsMSAAccount": false
            }
        },
        "Server": {
            "LicenseFilePath": "[License File Path - Server ]",
            "Installer": {
                "Path": "[Server Installer Path]",
                "InstallDir": "[Server Install Dir - Ex. C:\\ArcGIS\\Server]",
                "InstallDirPython": "[Python Install Dir - Ex. C:\\Python27] ",
                "EnableArcMapRuntime": "[ Optional - True (Default) | False - Enables ArcMapRuntime. Available starting 10.9.1]",
                "EnableDotnetSupport": "[ Optional - True | False (Default)  - Enables .Net Extension Support. Available starting 10.9.1]"
            },
            "ServerDirectoriesRootLocation": "\\\\[Fileshare Machine Name 1]\\[Fileshare Name]\\arcgisserver\\directories",
            "ConfigStoreLocation": "\\\\[Fileshare Machine Name 1]\\[Fileshare Name]\\arcgisserver\\config-store",
            "ExternalLoadBalancer": "alb.domain.com",
            "PrimarySiteAdmin": {
                "UserName": "[PrimarySiteAdmin Username]",
                "Password": "[PrimarySiteAdmin Password]"
            }
        },
        "Portal": {
            "LicenseFilePath": "[License File Path - Portal ]",
            "PortalLicenseUserTypeId": "[Portal License User Type Id- Required for 10.7 and above]",
            "Installer": {
                "Path": "[Portal Installer Path]",
                "WebStylesPath": "[Optional Parameter starting ArcGIS Enterprise 10.7.1 - Portal Web Styles Installer Path]",
                "InstallDir": "[Portal Install Dir - Ex. C:\\ArcGIS\\Portal]",
                "ContentDir": "[Portal Content Dir for Install- Ex. C:\\arcgisportal"
            },
            "ContentDirectoryLocation": "\\\\[File Share Machine 1]\\[File Share Name]\\arcgisportal\\content",
            "ExternalLoadBalancer": "alb.domain.com",
            "PortalAdministrator": {
                "UserName": "[PortalAdministrator Username]",
                "Email": "[PortalAdministrator Email]",
                "Password": "[PortalAdministrator Password]",
                "SecurityQuestionIndex": 1,
                "SecurityAnswer": "vanilla"
            }
        },
        "DataStore": {
            "ContentDirectoryLocation": "[Content Dir DataStore - C:\\arcgisdatastore]",
            "EnableFailoverOnPrimaryStop": false,
            "Installer": {
                "Path": "[DataStore Installer Path]",
                "InstallDir": "[DataStore Install Dir - Ex. C:\\ArcGIS\\DataStore]"
            }
        }
    }
}
```

Here is an additional resource that may be helpful when it comes to configuring the load balancer:

https://enterprise.arcgis.com/en/portal/latest/administer/windows/using-a-reverse-proxy-server-with-portal-for-arcgis.htm

**Layer 3/4 load balancer (Azure Load Balancer):**

> Note: When not using the ArcGIS Web Adaptor, the default context (/arcgis) must be used for the site. When integrating multiple Portal for ArcGIS and ArcGIS Server sites on the same layer 3/4 load balancer, a unique DNS record should be used for each site and Server Name Indication (SNI) used to route traffic to the appropriate back-end targets.

**Layer 7 load balancer (Azure Application Gateway Load Balancer):**

> Note: If you aren't using ArcGIS Web Adaptor with your portal, confirm that the Host header set by the load balancer matches the host name of the machine where Portal for ArcGIS is installed.

Thanks,
Cameron K.

**Reply by jesa-esrich** [reply]
@cameronkroeker Thanks for the detailed answer. We will try it this way and if we run into any problems open a new issue through the repo.
